### PR TITLE
Make FindSerializerForType faster

### PR DESCRIPTION
**Pull request by alexvaluyskiy**
Related https://github.com/akkadotnet/akka.net/issues/2566

I've reused some parts of Scala implementation
**Before**: 1 400 243 ops/s
**After**: 4 551 167 ops/s

### CODE REVIEW

**Comment by Horusiath** [comment]
Looks like DData types have been resolved to NewtonsoftJsonSerializer instead of their custom implementation (which is hyperion-based).

File: src/core/Akka/Serialization/Serialization.cs [hunk 1]
```diff
@@ -229,20 +229,34 @@ public Serializer FindSerializerFor(object obj)
...
-                return _serializerMap[_objectType];
+                //do a final check for the "object" serializer
+                if (_serializerMap.ContainsKey(_objectType) && _objectType.IsAssignableFrom(type))
+                    serializer = _serializerMap[_objectType];
 
-            throw new SerializationException($"Serializer not found for type {objectType.Name}");
+                if (serializer == null)
+                    throw new SerializationException($"Serializer not found for type {objectType.Name}");
+
+                _serializerMap.Add(type, serializer);
```
Comment (Rattenkrieg) [inline src/core/Akka/Serialization/Serialization.cs]: Shouldn't that become threadsafe collection in presence of updates? In Scala ConcurrentHashMap is used.

Comment (Rattenkrieg) [inline src/core/Akka/Serialization/Serialization.cs]: Also there will be duplicate exception if we first found object on line 251 and then trying to add it back. What the point of this btw?

Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serialization.cs]: We will gen an exception only if we remove all serializers from all configs

Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serialization.cs]: I've changed it to ConcurrentDictionary

Comment (Aaronontheweb) [inline src/core/Akka/Serialization/Serialization.cs]: Why not just change it to `_serializerMap.Add[type] = serializer;` ? Becomes an upsert that way.

File: src/core/Akka/Serialization/Serialization.cs [hunk 2]
```diff
@@ -229,20 +230,37 @@ public Serializer FindSerializerFor(object obj)
...
+                        serializer = serializerType.Value;
+                        break;
+                    }
+                }
 
-            //do a final check for the "object" serializer
-            if (_serializerMap.ContainsKey(_objectType) && _objectType.IsAssignableFrom(type))
-                return _serializerMap[_objectType];
+                //do a final check for the "object" serializer
+                if (serializer == null && _serializerMap.ContainsKey(_objectType) && _objectType.IsAssignableFrom(type))
```
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serialization.cs]: Why we need to check that type is subclass of objec?

Comment (Aaronontheweb) [inline src/core/Akka/Serialization/Serialization.cs]: lol.... it's literally checking to see if we're assignable from `object` ?

File: src/core/Akka/Serialization/Serialization.cs [hunk 3]
```diff
@@ -6,27 +6,23 @@
...
 using Akka.Util.Internal;
 
 namespace Akka.Serialization
 {
     /// <summary>
-    /// TBD
+    /// Serialization information needed for serializing local actor refs.
     /// </summary>
-    public class Information
+    internal class Information
```
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serialization.cs]: Made it internal

File: src/core/Akka/Serialization/Serialization.cs [hunk 4]
```diff
@@ -60,7 +56,7 @@ public static T SerializeWithTransport<T>(ActorSystem system, Address address, F
 
         private readonly Serializer _nullSerializer;
 
-        private readonly Dictionary<Type, Serializer> _serializerMap = new Dictionary<Type, Serializer>();
+        private readonly ConcurrentDictionary<Type, Serializer> _serializerMap = new ConcurrentDictionary<Type, Serializer>();
```
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serialization.cs]: Changed to ConcurrentDictionary

File: src/core/Akka/Serialization/Serialization.cs [hunk 5]
```diff
@@ -294,12 +300,7 @@ public static string SerializedActorPath(IActorRef actorRef)
             }
         }
 
-        /// <summary>
-        /// TBD
-        /// </summary>
-        /// <param name="serializerId">TBD</param>
-        /// <returns>TBD</returns>
-        public Serializer GetSerializerById(int serializerId)
+        internal Serializer GetSerializerById(int serializerId)
```
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serialization.cs]: This one should be internal

File: src/core/Akka/Serialization/Serializer.cs [hunk 6]
```diff
@@ -87,22 +88,11 @@ public byte[] ToBinaryWithAddress(Address address, object obj)
...
-        /// </remarks>
-        /// <param name="type">TBD</param>
-        /// <returns>TBD</returns>
-        protected static string TypeQualifiedNameForManifest(Type type)
-        {
-            return type == null ? string.Empty : string.Format("{0},{1}", type.FullName, type.Assembly.GetName().Name);
-        }
+        /// <param name="bytes">The array containing the serialized object</param>
+        /// <returns>The object contained in the array</returns>
+        public T FromBinary<T>(byte[] bytes) => (T)FromBinary(bytes, typeof(T));
```
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serializer.cs]: I think we should have this method

File: src/core/Akka/Serialization/Serializer.cs [hunk 7]
```diff
@@ -87,22 +88,11 @@ public byte[] ToBinaryWithAddress(Address address, object obj)
...
         /// </summary>
-        /// <remarks>
-        /// See <see cref="Type.GetType(string)"/> for details on how a type is looked up
-        /// from a name. In particular, if the (partial) assembly name is not included
-        /// only the assembly calling <see cref="Type.GetType(string)"/> is searched.
-        /// If the (partial) assembly name is included, it searches in the specified assembly.
-        /// </remarks>
-        /// <param name="type">TBD</param>
-        /// <returns>TBD</returns>
-        protected static string TypeQualifiedNameForManifest(Type type)
```
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serializer.cs]: We have an extension method for than. We don't need this method anymore

File: src/core/Akka/Serialization/Serialization.cs (hunk 1 above)
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/Serialization.cs]: changed

**Pull request merged**